framework/src/base/MooseApp.C

```diff
@@ -343,13 +344,29 @@ MooseApp::MooseApp(InputParameters parameters)
     _popped_final_mesh_generator(false)
 {
 #ifdef HAVE_GPERFTOOLS
+  if (std::getenv("MOOSE_PROFILE_BASE") && std::getenv("MOOSE_HEAP_BASE"))
+    mooseError("Can not do CPU and heap profiling together");
+
+  // For CPU profiling, users need to have envirement 'MOOSE_PROFILE_BASE'
   if (std::getenv("MOOSE_PROFILE_BASE"))
   {
     static std::string profile_file =
         std::getenv("MOOSE_PROFILE_BASE") + std::to_string(_comm->rank()) + ".prof";
-    _profiling = true;
+
+    _cpu_profiling = true;
     ProfilerStart(profile_file.c_str());
   }
+
+  // For Heap profiling, users need to have 'MOOSE_HEAP_BASE'
+  if (std::getenv("MOOSE_HEAP_BASE"))
+  {
+    static std::string profile_file =
+        std::getenv("MOOSE_HEAP_BASE") + std::to_string(_comm->rank()) + ".prof";
```

Comment:
I don't think we want the `.prof` string in here do we? At least on my machine, libtcmalloc actually appends a `.heap` suffix on its own

Reply:
Ok, I will remove that

Reply:
Done